implement legacy osu! cursor particles for Relax and Autopilot

Cai1Hsu opened the pull request:
In stable, with Relax enabled, cursor particles spawn when:
- hitting a circle or slider head with a random initial direction using the color of the hit circle or slider head
- sliding or spinning(even if you miss it) with the direction of `mouseVelocity` * 4 using the color of starBreakAdditive (like Kiai mode)

[reference](https://github.com/ppy/osu/issues/13760#issuecomment-907931613)

### Videos:
#### this PR:

https://github.com/ppy/osu/assets/47081443/6a35fbe3-e1b2-47f0-a798-985670148fd7


#### stable

https://github.com/ppy/osu/assets/47081443/0b7377a3-9330-4b35-a62d-81cbfe88ea25

Cai1Hsu committed:
make `LastParticleAdded` `CurrentIndex` `Particles` protected

Cai1Hsu committed:
implement legacy osu! cursor particles for Relax

Cai1Hsu committed:
Add test for relaxHitSpewer

Cai1Hsu committed:
rename `CreateSixParticles` to `SpawnRelaxHitParticles`

Cai1Hsu committed:
settle conflict

Cai1Hsu committed:
remove ZWNBSP

WitherFlower commented:
I'd like to mention that this used to be a feature in normal gameplay on old clients, where star particles would burst out from new combo circles during kiai. Not sure why it was removed, but this could be the occasion to add it back (even if only as a skinning option).

Cai1Hsu committed:
Fix particles spawn constantly

Cai1Hsu commented:
> where star particles would burst out from new combo circles during kiai.

That's already in the current version. This PR focus on the particles when you hit a circle with Relax, even if you are not under Kiai mode.

`if ((Relaxing || Relaxing2 || KiaiActive) && !EventManager.BreakMode)`

WitherFlower commented:
I think you misunderstood my comment, the code you sent only applies to sliders and spinners. I'm talking about the stars bursting out from circles. Here's a video with timestamp showing the old behaviour i'm referring to (nomod, 2012 client, during kiai) : https://youtu.be/meGiP1THDfw?t=64

bdach commented on osu.Game/Graphics/ParticleSpewer.cs:
```diff
@@ -171,7 +178,7 @@ protected override void Blit(IRenderer renderer)
                     );
 
-                    renderer.DrawQuad(Texture, quad, DrawColourInfo.Colour.MultiplyAlpha(alpha),
+                    renderer.DrawQuad(Texture, quad, p.Colour.MultiplyAlpha(alpha),
```
Not sure this is correct? This means that the spewer's colour is discarded entirely and the only colour that actually matters is the particle's?

Cai1Hsu replied:
at least it is so in the stable version. If we use spewer's color, there is a serious bug where all the particles' color changes when we want to switch colors for new particles.

bdach commented on osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs:
```diff
@@ -67,12 +74,26 @@ private void load(ISkinSource skin)
             if (player != null)
                 ((IBindable<bool>)breakSpewer.Active).BindTo(player.IsBreakTime);
+
+            relaxing = isRelaxing();
+
+            if (relaxing)
+                kiaiSpewer.Active.BindValueChanged(_ => { });
```
What is even happening here? Why does this conditionally register an _empty_ `BindValueChanged()` callback?

Cai1Hsu replied:
sorry for that, i forgot to clean it after debugging.

Cai1Hsu replied:
I tried `player.Mods.BindValueChanged(v => relaxing = v.NewValue.OfType<OsuModRelax>().Any());` and found this callback was never called. Maybe that's the reason.

bdach commented on osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs:
```diff
@@ -248,6 +347,23 @@ private enum SpewDirection
+                SpawnKiaiParticles = false;
+            }
+
+            public void SpawnNewParticles() => SpawnParticle();
```
Why is this named `SpawnNewParticles()` (plural) if it spawns just one? And why is this an actual method, which is next to `SpawnKiaiParticles` and also `CanSpawnParticles` (which this one method doesn't even check)?

Cai1Hsu replied:
`SpawnKiaiParticles` is used for the effect I said below, while `CanSpawnParticles` is used for controlling spawn particles in `Update()`. Considering `Update()` spawns constantly if `Active` and `CanSpawnParticles` were true, it's difficult to make hit particles spawn this way. So I exposed `spawnParticle()` to do so.

My previous code may be more reasonable.
``` c#
            public void SpawnRelaxHitParticles()
            {
                if (!base.CanSpawnParticles) return;

                for (int i = 0; i < 6; i++)
                    SpawnParticle();
            }
```

bdach commented on osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs:
```diff
@@ -84,6 +105,50 @@ protected override void Update()
+            }
+        }
+
+        private bool doRelaxingSpew()
+        {
+            if (playfield == null || gameplayState == null) return false;
+
+            bool hitting = playfield.HitObjectContainer.AliveObjects.FirstOrDefault(isHit) != null;
+            bool slidingOrSpinning = playfield.HitObjectContainer.AliveObjects.FirstOrDefault(isTracking) != null;
+
+            return hitting && relaxingKeyPressed() && !breakSpewer.Active.Value && !slidingOrSpinning;
+        }
+
+        private bool relaxingKeyPressed()
+        {
+            if (gameplayState == null) return false;
+
+            return relaxing && (leftPressed || rightPressed);
+        }
+
+        private bool isHit(DrawableHitObject h)
+        {
+            switch (h)
+            {
+                case DrawableSlider slider:
+                    return slider.HeadCircle.IsHit;
+
+                case DrawableHitCircle hitCircle:
+                    return hitCircle.IsHit;
+            }
+
+            return false;
```
These should probably be one method, because this logic is really stretched across too much

bdach commented on osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs:
```diff
@@ -67,12 +74,26 @@ private void load(ISkinSource skin)
+                kiaiSpewer.Active.BindValueChanged(_ => { });
         }
 
+        private bool isRelaxing() => player != null && player.Mods.Value.OfType<OsuModRelax>().Any();
```
Very rarely do we allow this sort of stuff where a drawable component queries active mods. Maybe fine in this case, not sure.

bdach commented on osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyCursorParticles.cs:
```diff
@@ -84,6 +105,50 @@ protected override void Update()
+
+            if (relaxing != relaxHitSpewer.Active.Value) relaxHitSpewer.Active.Value = relaxing;
+
+            if (relaxingKeyPressed() && !kiaiSpewer.Active.Value)
+            {
+                relaxHitSpewer.Direction = SpewDirection.None;
+                relaxHitSpewer.Colour = starBreakAdditive;
+                relaxHitSpewer.SpawnKiaiParticles = true;
+            }
+            else
+            {
+                relaxHitSpewer.SpawnKiaiParticles = false;
+            }
```
What is even happening here with the `SpawnKiaiParticles` thing. Why does the relax spewer spawn kiai particles _when the kiai spewer doesn't_

Cai1Hsu replied:
First, the reason why I don't use `kiaiSpewer` directly is that when `kiaiSpewer.Active` is `false` all the particles that were spawned would be hidden. So I use `relaxHitSpewer` to do his job. But when you are in Kiai mode and `relaxHitSpewer` is active, it would spawn twice more particles. This means I must disable either of them.

bdach commented:
Gonna be honest, I hate this feature and everything it stands for and would be in favour of not porting it. Implementation-wise it's so ugly and cross-cutting that I'm not sure there's even a clean way to implement it.

To the point that I'd probably much rather have this thing behind a skin toggle or setting, and have relax toggle that setting.

Cai1Hsu committed:
Fix particles were hidden immediately when disabling the spewer

Cai1Hsu committed:
1. code refactoring
2. Enable spawning on both Relax and Autopilot (Finally figure out Relaxing2 in the reference code is actually Autopilot)

Cai1Hsu committed:
add toggle
